### ISSUE TITLE
ReadObject with empty or small range results in error

### ISSUE THREAD

**Issue by shaileshkm** [googleapis/google-cloud-cpp#6516]
**Does this issue affect the google-cloud-cpp project?**
If the problem is with the Google Cloud service exposed by the `google-cloud-cpp` libraries instead of the client libraries themselves, you may consider opening a support request instead. The `google-cloud-cpp` developers cannot help you troubleshoot problems with the service itself.

Doing read_stream.status().message.data() returns
0x7fffeef27558 <std::string::_Rep::_S_empty_rep_storage+24> ""
I suspect this is SDK issue.

**What component of `google-cloud-cpp` is this related to?**
For example, is this related to bigtable (i.e., something in `google/cloud/bigtable`), or GCS (i.e., something in `google/cloud/storage`)?

Storage

**Describe the bug**

If I run ReadObject on a zero sized object, error is set on stream with empty buffer in status.

**To Reproduce**
1. Create empty/0 byte object.
2. `gcs::ObjectReadStream read_stream = client.ReadObject(_bucket_name.data(), _object_name.data(), gcs::ReadRange(start, start+readsize));
`
where readsize is 0.
3. Check read_stream && read_stream.IsOpen()
4. Obtain read_stream.status().message().data()

**Expected behavior**
My expectations are:
- If I'm able to create a 0 sized object in store, I should be able to read it. SDK minimally should check if the object is accessible.
- Even if it decides to set an error, it should be legible message.

**Operating system:**
If you are using a Linux distribution please include the name and version of the distribution too.
SUSE Linux Enterprise Server 12 SP4

**What compiler and version are you using?**
Please include the output of `g++ -v` or `clang++ -v` or the equivalent command-line flag.
cc version 9.2.1 20191018 

**What version of `google-cloud-cpp` are you using?**
Please include the output from `git rev-parse HEAD` if you are compiling from source, or the version number from the applicable `google/cloud/*/version.h` file.
1.24

**Additional context**

**Comment by coryan** [comment]
Thanks for the detailed bug report.  

> If I'm able to create a 0 sized object in store, I should be able to read it. SDK minimally should check if the object is accessible.
> Even if it decides to set an error, it should be legible message.

I think those things are true.  Unfortunately there is no way to express "read 0 bytes" in the JSON protocol, none that comes to mind immediately. You can read 1 byte, you can even read all the bytes (and it returns 0 bytes), but asking for exactly 0 bytes cannot be expressed.

That is to say: I think this is a valid bug report, but cannot figure out how to fix it right away.

**Comment by shaileshkm** [comment]
In that case, the error message could be improved. Currently, error is set on stream, but status().message has empty string.
In my code, I'm implementing a WA to get metadata and check size == 0. This will at least confirm access.

**Comment by coryan** [comment]
You could always try to read 1 byte, maybe the first byte.  That is about as good as reading 0 bytes performance-wise, and if there are 0 bytes it should work anyway.  In fact, I am temped to implement "read 0 bytes" as "read 1 byte but discard the first byte that arrives, if any".

**Comment by coryan** [comment]
Argh, you are correct:

```shell
curl -sSL -H "Range: bytes=0-0" -H "Authorization: Bearer $(gcloud auth print-access-token)" https://coryan-test-bucket.storage.googleapis.com/empty.txt
<?xml version='1.0' encoding='UTF-8'?><Error><Code>InvalidRange</Code><Message>The requested range cannot be satisfied.</Message><Details>bytes=0-0</Details></Error>
```

This works:

```shell
curl -sSL -H "Range: bytes=-1" -H "Authorization: Bearer $(gcloud auth print-access-token)" https://coryan-test-bucket.storage.googleapis.com/empty.txt
```

Which should be implementable using `gcs::ReadLast(1)` (I have not tested this):

https://googleapis.dev/cpp/google-cloud-storage/latest/structgoogle_1_1cloud_1_1storage_1_1v1_1_1ReadLast.html

It also seems like GCS supports the `Range: 1-0` convention:

```shell
curl -sSL -H "Range: bytes=1-0" -H "Authorization: Bearer $(gcloud auth print-access-token)" https://coryan-test-bucket.storage.googleapis.com/empty.txt
```

Seems like I could fix the client library to implement any empty range as `Range: bytes=1-0`... worth a try.

**Comment by shaileshkm** [comment]
For some reasons, for a small file (126 bytes), I'm getting same behavior as 0 sized file.
Read call (with Range(0,126)) fails with message set to
0x7fffeef27558 std::string::_Rep::_S_empty_rep_storage+24 ""

For a large file (7261125) I see no issue.

Any idea?

**Comment by coryan** [comment]
Sorry for the delay, I just tested reading small objects with the client library and it works for me (I will submit a PR with this integration test later today I hope):

First create an object of 446 bytes:

```
2021-05-12T20:24:37.641750230Z [INFO] <140334594172928> InsertObjectMedia() >> payload={ObjectMetadata={name=ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt, acl=[], bucket=coryan-test-bucket, cache_control=, component_count=0, content_disposition=, content_encoding=, content_language=, content_type=application/octet-stream, crc32c=6Y46Mg==, etag=COa81uX8xPACEAE=, event_based_hold=false, generation=1620851077520998, id=coryan-test-bucket/ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt/1620851077520998, kind=storage#object, kms_key_name=, md5_hash=96HF9K981B+JfoQuTVnyCg==, media_link=https://storage.googleapis.com/download/storage/v1/b/coryan-test-bucket/o/ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt?generation=1620851077520998&alt=media, metageneration=1, name=ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt, retention_expiration_time=1970-01-01T00:00:00Z, self_link=https://www.googleapis.com/storage/v1/b/coryan-test-bucket/o/ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt, size=446, storage_class=MULTI_REGIONAL, temporary_hold=false, time_created=1620851077598000000, time_deleted=0, time_storage_class_updated=1620851077598000000, updated=1620851077598000000}} (../google/cloud/storage/internal/logging_client.cc:49)
```

Reading 1 byte works:

```
2021-05-12T20:24:37.641917039Z [INFO] <140334594172928> ReadObject() << ReadObjectRangeRequest={bucket_name=coryan-test-bucket, object_name=ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt, disable-md5-hash=1, read-range=ReadRangeData={begin=0, end=1}} (../google/cloud/storage/internal/logging_client.cc:75)
2021-05-12T20:24:37.642375361Z [INFO] <140334594172928> Read() current_offset=0 (../google/cloud/storage/internal/retry_object_read_source.cc:49)
```

And a few bytes in the middle:

```
2021-05-12T20:24:37.699548084Z [INFO] <140334594172928> ReadObject() << ReadObjectRangeRequest={bucket_name=coryan-test-bucket, object_name=ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt, disable-md5-hash=1, read-range=ReadRangeData={begin=4, end=8}} (../google/cloud/storage/internal/logging_client.cc:75)
2021-05-12T20:24:37.699734804Z [INFO] <140334594172928> Read() current_offset=4 (../google/cloud/storage/internal/retry_object_read_source.cc:49)
```

and all the bytes using `ReadRange(0, 446)`:

```
2021-05-12T20:24:37.734366015Z [INFO] <140334594172928> ReadObject() << ReadObjectRangeRequest={bucket_name=coryan-test-bucket, object_name=ob-R4tMS4Xji3LERHlR44Xq4PjznUs3ob0xmRXvxuEHxiobfRZ3D06xTzfbkypS1YVamKguBaQ0lrX5GWjflueTSQdyxPYhu6if0QufgZh3E1kPuzztcr1XblDKLph69vZp.txt, disable-md5-hash=1, read-range=ReadRangeData={begin=0, end=446}} (../google/cloud/storage/internal/logging_client.cc:75)
2021-05-12T20:24:37.734579165Z [INFO] <140334594172928> Read() current_offset=0 (../google/cloud/storage/internal/retry_object_read_source.cc:49)
```



PS: the issue around reading 0 bytes is still there.

**Comment by coryan** [comment]
BTW, `Read: 1-0` "works" but returns all the bytes, definitely not what we want.

**Comment by shaileshkm** [comment]
Ok. I will try again maybe tomorrow. 
Thanks for confirming.
Is there a way to turn on tracing for client library?

**Comment by coryan** [comment]
> Is there a way to turn on tracing for client library?

You need to set two environment variables (or make some function calls):

```shell
export GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes
export CLOUD_STORAGE_ENABLE_TRACING=raw-client
```

This is documented in:

https://googleapis.dev/cpp/google-cloud-storage/latest/

**Comment by shaileshkm** [comment]
Hi Carlos,
Sorry I couldn't get back to you earlier.
I'm still facing trouble when I try read small file (126 bytes).
Here are the trace messages. Can you tell me how I how I should debug this?

2021-05-14T18:30:18.564301294Z [INFO] <139939090360576> Enabling logging for RawClient functions (/home/I530084/gcs/google-cloud-cpp/google/cloud/storage/client_options.cc:159)
Setting valid connection? success
2021-05-14T18:30:18.564786860Z [INFO] <139939090360576> GetObjectMetadata() << GetObjectMetadataRequest={bucket_name=basic_test3, object_name=encrypt-load-single-ascii-smallint.inp} (/home/I530084/gcs/google-cloud-cpp/google/cloud/storage/internal/logging_client.cc:46)
2021-05-14T18:30:18.711268192Z [INFO] <139939090360576> GetObjectMetadata() >> payload={ObjectMetadata={name=encrypt-load-single-ascii-smallint.inp, acl=[], bucket=basic_test3, cache_control=, component_count=0, content_disposition=, content_encoding=, content_language=, content_type=, crc32c=ci1m9w==, etag=CMy2kPzmyfACEAE=, event_based_hold=false, generation=1621017017719628, id=basic_test3/encrypt-load-single-ascii-smallint.inp/1621017017719628, kind=storage#object, kms_key_name=, md5_hash=rp4PJZUX/2yiXIhTaRVzBg==, media_link=https://storage.googleapis.com/download/storage/v1/b/basic_test3/o/encrypt-load-single-ascii-smallint.inp?generation=1621017017719628&alt=media, metageneration=1, name=encrypt-load-single-ascii-smallint.inp, retention_expiration_time=1970-01-01T00:00:00Z, self_link=https://www.googleapis.com/storage/v1/b/basic_test3/o/encrypt-load-single-ascii-smallint.inp, size=126, storage_class=STANDARD, temporary_hold=false, time_created=1621017017829000000, time_deleted=0, time_storage_class_updated=1621017017829000000, updated=1621017017829000000}} (/home/I530084/gcs/google-cloud-cpp/google/cloud/storage/internal/logging_client.cc:49)
Going to read from 0, end 125
2021-05-14T18:30:18.711629582Z [INFO] <139939090360576> ReadObject() << ReadObjectRangeRequest={bucket_name=basic_test3, object_name=encrypt-load-single-ascii-smallint.inp, disable-md5-hash=1, read-range=ReadRangeData={begin=0, end=126}} (/home/I530084/gcs/google-cloud-cpp/google/cloud/storage/internal/logging_client.cc:75)
2021-05-14T18:30:18.712988497Z [INFO] <139939090360576> Read() current_offset=0 (/home/I530084/gcs/google-cloud-cpp/google/cloud/storage/internal/retry_object_read_source.cc:49)

**Comment by coryan** [comment]
That looks like a successful trace to me.... Are there any error messages?  Did you check `.status()` on the stream after reading?  Did you look at the integration test in #6551?  Should I be writing the test any different so it is more similar to your application?

Meanwhile, I will try to change integration test to write exactly 126 bytes and see if there are any problems.

**Comment by shaileshkm** [comment]
After invoking ReadObject, I do following to determine to check for failure:
```
if(read_stream && read_stream.IsOpen()) {
   // success
   return(read_stream.gcount();
}
// failure
ProcessError(read_stream.status().message().data());

```
The error is as reported in the first post.
0x7fffeef27558 std::string::_Rep::_S_empty_rep_storage+24 ""

**Comment by coryan** [comment]
Sorry for the delay, I was trying to finish other things.

I wrote a test (see link below) to reproduce the problem, without success.  Could you examine the test and let me know if I should be doing something different to reproduce your use case?

https://github.com/googleapis/google-cloud-cpp/compare/main...coryan:test-reproduce-6516

I am testing with openSUSE next, in case it turns out to be compiler or library specific, but I think that unlikely.

**Comment by coryan** [comment]
I added a test that does not read the data, and compiled on openSUSE, I am not that familiar with openSUSE vs. SLES equivalency, so maybe you can advice me if this is a useful comparison or not:

```
cat /etc/os-release 
NAME="openSUSE Leap"
VERSION="15.2"
ID="opensuse-leap"
ID_LIKE="suse opensuse"
VERSION_ID="15.2"
PRETTY_NAME="openSUSE Leap 15.2"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:leap:15.2"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
```

**Comment by shaileshkm** [comment]
Thanks Carlos.

I ran a very simple test, which creates a client, gets size, reads the file and finishes.
That works.
This leads me to believe, it has something to do with the way I'm creating/assigning client and leaks I had reported earlier.
In the test that is failing, I create client multiple times (in order to do positive/negative tests). I haven't figured out clean way of freeing earlier client before I assign new one.
Here's how I'm doing it.

I have my own Client class which aggregates
`google::cloud::StatusOr<gcs::Client>    _client_wrapper;`

I first authenticate using Service Account 
```
- Invalidate previously authenticated client by doing following
   _client_wrapper = google::cloud::StatusOr<gcs::Client>();
- Get credential object using CreateServiceAccount... method
- _client_wrapper = gcs::Client(gcs::ClientOptions(*creds));
- if( ! _client_wrapper) do error reporting
```
However, _client_wrapper object stays as it is till the time, I try to authenticate again for next set of tests.

Once authenticated successfully, I use 
`gcs::Client& client = _client_wrapper.value();`
to invoke operations like ReadObject.

Is there a better way to 'release' previous client object before I authenticate again?

**Comment by coryan** [comment]
> 
> I ran a very simple test, which creates a client, gets size, reads the file and finishes.
> That works.

Good.

> This leads me to believe, it has something to do with the way I'm creating/assigning client and leaks I had reported earlier.

I am not sure why a leak would explain a crash, not until the memory is exhausted.  

> In the test that is failing, I create client multiple times (in order to do positive/negative tests).

Okay, do you do this in a single thread?  Are there multiple threads acting on this client?

> I haven't figured out clean way of freeing earlier client before I assign new one.
> Here's how I'm doing it.
> 
> I have my own Client class which aggregates
> `google::cloud::StatusOr<gcs::Client> _client_wrapper;`
> 
> I first authenticate using Service Account

A small quibble: authentication does not happen until you actually make a request.  You probably mean: "load the credentials using Service Account".  The distinction is important, loading the credentials can be successful, and then authentication (or authorization) may fail.

> ```
> - Invalidate previously authenticated client by doing following
>    _client_wrapper = google::cloud::StatusOr<gcs::Client>();

Note that this client is not "usable": calling `.value()` throws an exception, and `operator*` is undefined behavior.

> - Get credential object using CreateServiceAccount... method

Presumably you check those for an error?  What do you do in that case?

> - _client_wrapper = gcs::Client(gcs::ClientOptions(*creds));
> - if( ! _client_wrapper) do error reporting

At this point that error reporting will never happen.  That is because you are initializating `gcs::StatusOr<gcs::Client>` with the "success" case: an object of type `gcs::Client`.  If you wanted to check for errors you should have done this before calling `*creds`?

What does "do error reporting" entails?  Does the code continue to run?  If so, do you check `_client_wrapper.ok()` before each use in your class?

> ```
> 
> However, _client_wrapper object stays as it is till the time, I try to authenticate again for next set of tests.
> 
> Once authenticated successfully, I use
> `gcs::Client& client = _client_wrapper.value();`
> to invoke operations like ReadObject.

I would write that as `gcs::Client client = _client_wrapper.value();` no need to use a reference.

> Is there a better way to 'release' previous client object before I authenticate again?

"Better" requires me to understand the tradeoffs in your application, which I don't.  Modulo the questions above, what you are doing is largely Okay.  I would not necessarily write it as you did, but there is nothing fundamentally wrong with it.  The reason I say I would not write it as you did is because there is little point in trying to reload the credentials (assuming you are trying to reload them) after they fail, the errors you get in that case are things like "the file is not there", or "the format of the file is wrong".  Reloading is hardly ever going to fix that, but again, maybe it does in your application (maybe some external mechanism changes the file?)

I think it is more likely that the result from `ReadObject` is going out of scope or somehow getting deleted and you continue to use it.  Did you run your code with `valgrind(1)` or the address sanitizer?

**Comment by shaileshkm** [comment]
The unit test code is single threaded (which is facing this issue with small file read failing). I will test my server side code with the same file. The server side code is multi-threaded but it has better life cycle management of client (in the sense that _client_wrapper itself gets deleted).

> If you wanted to check for errors you should have done this before calling *creds?
Not sure what you meant here, but let me check for error at that stage as well.

I'm pretty sure, for the behavior wrt small file, there is no issue with credentials or client object validity.
Just before invoking ReadObject, I invoke GetObjectMetadata to determine size. I always get correct size (126).

What I'm suspecting is, one of previous negative tests state lingering around in client object.
Next I will try recompile SDK with debug flag. In server code execution I saw a crash in client destructor.

**Comment by shaileshkm** [comment]
I just checked my code, I do check cred object is valid immediately after invoking CreateServiceAccountCredentials API.

**Comment by shaileshkm** [comment]
Hi Carlos,

This is getting weird.

I began debugging my server side code. It has better life cycle management of client object, in the sense that it gets deleted after every import/export.

I had to recompile the SDK as statically linked library since our code traps new/delete and that caused trouble when linking with SDK .so.

Currently testing with a file of size 988890 bytes, just to make sure the file is not too small.
The Read logic is as below
```
Read(start, readsize, buffer) {
    stream = client.ReadObject(ReadRange(start, start+readsize));
    if(stream && stream.IsOpen()) {
        stream.read(buffer, readsize);
        if(stream.good()) {
           bytesread = stream.gcount();
           return(bytesread);
        }
    }

    ProcessError(stream.status().messate().data());
    return(-1);
}
```
The server logic always finds file size before doing any Read operation.

- First thing I discovered was ReadObject did not like range beyond the end of file. The code invoked it on range 0-1048576. Stream was opened successfully, but after doing read (of 1MB), stream.good() returned failure. Error message was same as earlier `<std::string::_Rep::_S_empty_rep_storage+24> ""`.
- I then modified caller code to ensure both `start` and `start+readsize` will always be <= size of file. The first read attempt was successful.
- Our server code has some peculiar logic which needs to reread few bytes from the end of file. For the call `Read(983040, 5850, buffer)` the ReadObject failed (stream.IsOpen() returned false). Error again is `<std::string::_Rep::_S_empty_rep_storage+24> ""`. This may be the same issue I reported earlier with small sized read.

This is very mystifying.

Another point. I'm testing with 1.25 Yesterday I tried to build with 1.27 but ran in compiler error storm with my server code, so went back to using 1.25.

**Comment by coryan** [comment]
In the code snippet: is `stream = client.ReadObject(...);` a member variable? A local variable?

Would it be possible to print `stream.rdbuf()`?  Is that a `nullptr`?

https://en.cppreference.com/w/cpp/io/basic_ios/rdbuf

**Comment by shaileshkm** [comment]
It's a local variable.
Thanks for the pointer.
I tried this after doing ReadObject:
```
    {   
        if( read_stream ) {
            auto rdbuf = read_stream.rdbuf();
            if( rdbuf ) {
                auto avail = rdbuf->in_avail();
                printf("Available: %d\n", avail);
            }
        }
    }
```
For the invocation `Read(983040, 5850, buffer)`
I get 
Available: 5850
The buffer returned by `rdbuf->gptr()` holds expected data.

What does this mean?

**Comment by shaileshkm** [comment]
Seems like, upon reaching EoF, 'Open' state is reset. That explains earlier observation with small file.
That's why my check was failing.

Is checking for non-null read_stream itself sufficient or I should check for non-null rdbuf?
In other words, which of following do you recommend?
```
if(read_stream && read_stream.rdbuf())
or
if(read_stream)
```

**Comment by coryan** [comment]
> Seems like, upon reaching EoF, 'Open' state is reset. That explains earlier observation with small file.
> That's why my check was failing.

Okay.

> Is checking for non-null read_stream itself sufficient

A small nit: the `read_stream` is never "null", it is not a pointer, it is a value object.  Doing `if (read_stream)` is actually calling `operator bool`:

https://en.cppreference.com/w/cpp/io/basic_ios/operator_bool

as you can see from the table at the bottom this operator returns `false` when `fail()` returns true, which is basically when either `failbit` or `badbit` are set.  The often confusing part (and you may know this already) is that reading N bytes when less than N bytes are available sets the `failbit`.

 or I should check for non-null rdbuf?
> In other words, which of following do you recommend?
> 
> ```
> if(read_stream && read_stream.rdbuf())
> or
> if(read_stream)
> ```

The second I think.

> I imagine, doing read_stream.read itself must be checking for rdbuf if NULL.

Yes.

**Comment by coryan** [comment]
I think the questions in this thread have been answered.  If you think there is something left for us to do please reopen (or create a more specific bug, whatever you prefer / it is easier for you).
